engine/consensus/dkg/reactor_engine.go

```diff
@@ -229,42 +229,59 @@ func (e *ReactorEngine) startDKGForEpoch(currentEpochCounter uint64, first *flow
 	e.registerPhaseTransition(curDKGInfo.phase3FinalView, dkgmodule.Phase3, e.end(nextEpochCounter))
 }
 
-// handleEpochCommittedPhaseStarted checks that the local DKG completed and
-// that our locally computed key share is consistent with the canonical key vector.
+// handleEpochCommittedPhaseStarted is invoked upon the transition to the EpochCommitted
+// phase, when the canonical beacon key vector is incorporated into the protocol state.
 //
-// If our key is inconsistent, we log a warning message.
-// CAUTION: we will still sign using the inconsistent key, to be fixed with Crypto
-// v2 by falling back to staking signatures in this case.
+// This function checks that the local DKG completed and that our locally computed
+// key share is consistent with the canonical key vector. When this function returns,
+// an end state for the just-completed DKG is guaranteed to be stored (if not, the
+// program will crash). Since this function is invoked synchronously before the end
+// of the current epoch, this guarantees that when we reach the end of the current epoch
+// we will either have a usable beacon key (successful DKG) or a DKG failure end state
+// stored, so we can safely fall back to using our staking key.
 //
-func (e *ReactorEngine) handleEpochCommittedPhaseStarted(currentEpochCounter uint64) {
+// CAUTION: This function is not safe for concurrent use. This is not enforced within
+// the ReactorEngine - instead we rely on the protocol event emission being single-threaded
+//
+func (e *ReactorEngine) handleEpochCommittedPhaseStarted(currentEpochCounter uint64, firstBlock *flow.Header) {
```

Comment:
During the last discussion we considered the option of passing in the Badger transaction context through the `EpochCommittedPhaseStarted` protocol event callback. I decided against doing this for two reasons:
* Since epoch-related protocol events are emitted when the relevant block is **finalized**, the block (and corresponding snapshot) are already accessible in the protocol state, and are immutable so access them here is safe.
* Many of the storage operations we need to perform here are targeting `DKGState`, which uses a separate physical database backend (`SecretsDB`), so we would be unable to have ACID semantics across all the storage operations here anyway.